### PR TITLE
Dynamic topo neighbor allreduce

### PR DESCRIPTION
**Pull request by hanbinhu**

### CODE REVIEW

File: bluefog/common/mpi_controller.cc [hunk 1]
```diff
@@ -291,24 +296,104 @@ void MPIController::NeighborAllreduce(TensorTableEntry& entry) {
...
+    for (int i = 0; i < nsend; ++i) {
+      int ret_code = MPI_Isend(sendbuf, num_elements, mpi_ctx_.GetMPIDataType(entry.tensor),
+          entry.send_neighbors->at(i), 0, mpi_ctx_.GetMPICommunicator(Communicator::GLOBAL),
+          &requests[i]);
+      if (ret_code != MPI_SUCCESS) {
+        throw std::runtime_error(
+            "MPI_Isend (for dynamic neighbor_allreduce) failed, see MPI output for details.");
+      }
+    }
+    MPI_Waitall(nsend+nrecv, requests.data(), statuses.data());
```
Comment (ybc1991) [inline bluefog/common/mpi_controller.cc:386]: Check the statuses. If not all succeed, callback sends error as well.

File: test/torch_ops_test.py [hunk 2]
```diff
@@ -297,6 +298,99 @@ def test_neighbor_allreduce_avg_precision(self):
...
+        self_weight = 0.0
+        neighbor_weights = {(rank-1) % size : 1.0}
+        send_ranks = [(rank + 2) % size]
+
+        dims = [1, 2, 3]
+        for dtype, dim in itertools.product(dtypes, dims):
+            tensor = torch.FloatTensor(*([23] * dim)).fill_(1).mul_(rank)
+            tensor = self.cast_and_place(tensor, dtype)
+            name = "neighbor_allreduce_{}_{}".format(dim, dtype)
+            with pytest.raises(Exception):
```
Comment (ybc1991) [inline test/torch_ops_test.py:323]: Narrow down the exception caught. It should be ValueError only.

File: bluefog/common/mpi_controller.cc [hunk 3]
```diff
@@ -273,15 +273,20 @@ void MPIController::NeighborAllreduce(TensorTableEntry& entry) {
   // Allgather output will have shape of:
   // (sum of first dimension of every tensor) x (tensor slice shape).
   // For allreduce, the first dimension of every tensor should be the same.
-  int total_entry_dimension_size =
-      entry.tensor->shape().dim_size(0) * GetNeighborSize();
   TensorShape output_shape;
+  int total_entry_dimension_size;
```
Comment (ybc1991) [inline bluefog/common/mpi_controller.cc:277]: Write like 
const int neighbor_size = entry.send_neighbors->empty() ? GetNeighborSize() : entry.recv_neighbors->size();
const int total_entry_dimension_size =  entry.tensor->shape().dim_size(0) * neighbor_size;

File: bluefog/common/mpi_controller.cc [hunk 4]
```diff
@@ -291,24 +296,104 @@ void MPIController::NeighborAllreduce(TensorTableEntry& entry) {
...
+    int ret_code = MPI_Allgather(send_check_buf, size*2, MPI_C_BOOL,
+                                 recv_check_buf, size*2, MPI_C_BOOL,
+                                 mpi_ctx_.GetMPICommunicator(Communicator::GLOBAL));
+    if (ret_code != MPI_SUCCESS) {
+        throw std::runtime_error(
+            "MPI_Allgather (for dynamic neighbor_allreduce negotiation) failed, see MPI output "
+            "for details.");
+    }
+    auto get_index = [size](int i, int j) -> int { return 2*size*i+j; };
+    for (int i = 0; i < size; ++i) {
```
Comment (ybc1991) [inline bluefog/common/mpi_controller.cc:321]: I think the check is equivalent to 
Recv = [2\*sizes\*size] => two  boolean size*size matrix. One for recieving pattern and another for sending pattern. The rule is the transpose of recieving matrix should be the same for sending one, right?

File: bluefog/torch/mpi_ops.cc [hunk 5]
```diff
@@ -312,17 +317,23 @@ int DoNeighborAllreduce(::torch::Tensor tensor, ::torch::Tensor output,
...
+    auto bf_recv_neighbors = std::make_shared<std::vector<int>>(recv_neighbors);
+    auto bf_send_neighbors = std::make_shared<std::vector<int>>(send_neighbors);
     auto ready_event = RecordReadyEvent(device);
     auto enqueue_result = EnqueueTensorNeighborAllreduce(
-        bf_context, bf_tensor, bf_output, ready_event, op_name, CPU_DEVICE_ID,
-        [handle, self_weight, neighbor_weights, avg_computation, cpu_output, tensor,
-         output, device, op_name, tid, timeline_ptr](const Status& status) mutable {
+        bf_context, bf_tensor, bf_output, ready_event, bf_recv_neighbors, bf_send_neighbors,
+        op_name, CPU_DEVICE_ID, [handle, self_weight, neighbor_weights, avg_computation, cpu_output,
+        tensor, recv_neighbors, send_neighbors, output, device, op_name, tid, timeline_ptr]
```
[the remark this answers is not in the document]
Comment (hanbinhu) [reply bluefog/torch/mpi_ops.cc:326]: Will be resolved in future refactoring

File: bluefog/torch/mpi_ops.cc [hunk 6]
```diff
@@ -312,17 +317,23 @@ int DoNeighborAllreduce(::torch::Tensor tensor, ::torch::Tensor output,
...
+        (const Status& status) mutable {
```
[the remark this answers is not in the document]
Comment (hanbinhu) [reply bluefog/torch/mpi_ops.cc:327]: Will be resolved in future refactoring

File: bluefog/common/mpi_controller.cc [hunk 7]
```diff
@@ -262,6 +262,65 @@ void MPIController::NeighborAllgather(TensorTableEntry& entry) {
...
+        throw std::runtime_error(
+            "MPI_Allgather (for dynamic neighbor_allreduce negotiation) failed, see MPI output "
+            "for details.");
+    }
+    auto GetSendIndex = [size](int i, int j) -> int { return 2*size*i+j; };
+    auto GetRecvIndex = [size](int i, int j) -> int { return 2*size*i+j+size; };
+    for (int i = 0; i < size; ++i) {
+        if (res) break;
+        for (int j = 0; j < size; ++j) {
+            if (recv_check_buf[GetSendIndex(i, j)] == recv_check_buf[GetRecvIndex(j, i)]) continue;
```
Comment (ybc1991) [inline bluefog/common/mpi_controller.cc:290]: 为啥不直接写  if(a!=b) { res=True; break; }

File: bluefog/torch/mpi_ops.py [hunk 8]
```diff
@@ -439,14 +442,15 @@ def neighbor_allreduce(tensor: torch.Tensor,
...
-                                            send_neighbors, name)
+                                            send_neighbors, enable_topo_check, name)
     return synchronize(handle)
 
 
 def neighbor_allreduce_nonblocking(tensor: torch.Tensor,
                                    self_weight: float = None,
                                    neighbor_weights: Dict[int, float] = None,
                                    send_neighbors: List[int] = None,
+                                   enable_topo_check: bool = False,
```
Comment (ybc1991) [inline bluefog/torch/mpi_ops.py:453]: Default 改成true吧
因为我们测performance 或者high level的optimizer可以关掉他。不会用的人让他能主动fail比较靠谱

File: bluefog/torch/mpi_ops.py [hunk 9]
```diff
@@ -461,16 +465,19 @@ def neighbor_allreduce_nonblocking(tensor: torch.Tensor,
...
             (in-)neighbors.
-        send_neighbors: the list of neighbor nodes to be sent to. If set to be None, assume the
-            topology is static. If having values, assume the topology is dynamic, this node will
-            only send its value to the neighbors listed in this variable.
+        send_neighbors: The list of neighbor nodes to be sent to. If set to be None, assume the
+            the current node sends to all of its (out-)neighbors. If having values, assume only
+            part of (out-)neighbors will be sent to. In this mode, this node sends its value to
+            partial neighbors listed in this variable.
+        enable_topo_check: When send_neighbors is present, enabling this option checks if the
+            sending and recieving neighbors match with each other.
```
Comment (ybc1991) [inline bluefog/torch/mpi_ops.py:480]: \+ Disabling this check can boost the performance.

**Pull request merged**